src/test/java/org/cactoos/iterator/NoNullsTest.java

```diff
@@ -23,72 +23,53 @@
  */
 package org.cactoos.iterator;
 
-import java.util.Iterator;
-import org.junit.Rule;
 import org.junit.Test;
-import org.junit.rules.ExpectedException;
 import org.llorllale.cactoos.matchers.Assertion;
+import org.llorllale.cactoos.matchers.MatcherOf;
 import org.llorllale.cactoos.matchers.Throws;
 
 /**
  * Test cases for {@link NoNulls}.
  *
  * <p>There is no thread-safety guarantee.
- *
- * @since 0.35
  * @checkstyle JavadocMethodCheck (500 lines)
+ * @since 0.35
  */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
 public final class NoNullsTest {
 
-    /**
-     * A rule for handling an exception.
-     */
-    @Rule
-    public final ExpectedException exception = ExpectedException.none();
-
     @Test
     public void nextThrowsErrorIfNull() {
         new Assertion<>(
             "Must throw exception",
             () -> new NoNulls<>(
-                new Iterator<Integer>() {
-                    @Override
-                    public boolean hasNext() {
-                        return true;
-                    }
-
-                    @Override
-                    public Integer next() {
-                        return null;
-                    }
-
-                    @Override
-                    public String toString() {
-                        return "Iterator@NoNullsTest";
-                    }
-                }
+                new IteratorOf<>(new String[]{null})
             ).next(),
             new Throws<>(
-                "Item #0 of Iterator@NoNullsTest is NULL",
+                new MatcherOf<>(
+                    (String msg) -> msg.matches("^Item #0 of .*? is NULL")
+                ),
                 IllegalStateException.class
             )
         ).affirm();
     }
 
-    /*
-    * @todo #1039:15min Currently it's impossible to match error messages
-    *  by a pattern or partially. Replace {@link Rule} with {@link Throws} after
-    *  <a href="https://github.com/llorllale/cactoos-matchers/issues/108">llorllale/cactoos-matchers#108</a>
-    *  is fixed
-    */
     @Test
     public void nthThrowsErrorIfNull() {
-        this.exception.expect(IllegalStateException.class);
-        final Iterator<String> itr = new NoNulls<>(
-            new IteratorOf<>("a", "b", null, "c")
-        );
-        while (itr.hasNext()) {
-            itr.next();
-        }
+        new Assertion<>(
+            "Must throw exception",
+            () -> new TailOf<>(
+                1,
+                new NoNulls<>(
+                    new IteratorOf<>("a", "b", null, "c")
+                )
+            ).next(),
```

Comment:
@scristalli I think it is the defect of the implementation of `TailOf`, in that it has [logic executed in constructor](https://www.yegor256.com/2015/05/07/ctors-must-be-code-free.html), and not of this test.

I also think the semantics is preserved, only it's expressed in a declarative form. Originally it was "iterate through all elements of the `Iterable` and expect an exception". Now it's "last element of the `Iterable` produces an exception". Since the only way to get the last element of an `Iterable` is by iterating through all of them, this is equivalent to "iterating through all elements of the `Iterable` produces an exception". It seems to me using some mapping object  instead just because it currently has a better implementation is not worth it.

Reply:
@vzurauskas I disagree. Your explanation is correct, but IMHO it clearly shows that an _implicit_ layer has been added, i.e. 

> the only way to get the last element of an Iterable is by iterating through all of them

Now, reading the code implies not only knowing this fact, but also noticing it immediately. Therefore, I think that a map operation of `next()` on the collection is better, and not only depending on current implementation.

Reply:
@scristalli It's not a collection, it's an iterable. You can't apply a map operation of `next()` on its elements because the elements are not of type `Iterator`. You could use identity, but that doesn't imply mapping would be applied immediately and not when actual iterating begins.

Reply:
@vzurauskas correct, my fault for not explaining correctly. I was thinking of something roughly equivalent to iteration, a more evident translation of the `for`, so anything is fine in that direction. If it can be done in EO paradigm, I'd advise we use it instead of relying on other internal logic. If not, let's keep your solution.

Reply:
@scristalli Yes, I wanted some object that would represent the result of iterating through all elements, something like [ReadBytes](https://github.com/yegor256/cactoos-http/blob/master/src/test/java/org/cactoos/http/io/ReadBytes.java), which represents the result of reading of all bytes from `InputStream`. `TailOf` seems to be the closest thing to this concept.